### PR TITLE
Ben/invert

### PR DESCRIPTION
**Pull request by blugavere**

### CODE REVIEW

**Comment by bryanmacfarlane** [comment]
Thanks.  Just so we're on the same page, can you add CR comments (not code) around most of your changes?  I'll review tomorrow

File: api/authrouter.ts [hunk 1]
```diff
@@ -35,7 +35,7 @@ export class AuthRouter {
             let decoded: any;
 
             try {
-                let auth: string = req.headers['authorization'];
+                let auth: string = <string>req.headers['authorization'];
```
Comment (blugavere) [inline api/authrouter.ts:38]: tsc was failing

File: api/identity.ts [hunk 2]
```diff
@@ -15,11 +14,11 @@ export interface Identity {
 }
     
 export class IdentityService {
-    constructor() {
-        this._store = new sm.Store();
+    constructor(private store: IStore) {
+        this._store = store;
```
Comment (blugavere) [inline api/identity.ts:18]: a repository abstraction is being used, but by referencing a concrete implementation explicitly (a store, directly referencing nedb) in its constructor theres tight coupling between business domain and the persistence layer.

injecting a store-like instance allows the persistence layer to be stubbed out for testability / flexibility. can change the store w/o modifying the service.

File: api/routes.ts [hunk 3]
```diff
@@ -25,10 +27,14 @@ export async function setup(app:exp.Express) {
     //----------------------------------------------------------
     // Create Services
     //----------------------------------------------------------
-    let idsvc: id.IdentityService = new id.IdentityService();
+    const idDb = new nedb();
+    const idStore = new sm.Store(idDb);
```
Comment (blugavere) [inline api/routes.ts:31]: moving the dependency construction up the stack to the routing layer keeps the infrastructure an app level concern, insulating business logic from infrastructure choices. as long as the store returns a promise that resolves to an array of t, the persistence implementation details can be ignored.

File: api/store.ts [hunk 4]
```diff
@@ -2,10 +2,17 @@ import * as path from 'path';
 import * as nedb from 'nedb';
 import * as Q from 'q';
 
-export class Store {
-    constructor() {
+export interface IStore {
```
Comment (blugavere) [inline api/store.ts:5]: program to an interface instead of a concrete implementation to provide flexibility

File: api/routes.ts [hunk 5]
```diff
@@ -25,10 +27,14 @@ export async function setup(app:exp.Express) {
...
+    let idsvc: id.IdentityService = new id.IdentityService(idStore);
     await idsvc.initialize();
 
-    let qsvc: qm.QuotesService = new qm.QuotesService();
+    const quoteDb = new nedb();
+    const quoteStore = new sm.Store(quoteDb);
```
Comment (blugavere) [inline api/routes.ts:36]: for large apps where this gets un-wieldy, you can use a DI container to automatically wire up all the deps.

**Comment by bryanmacfarlane** [comment]
Thanks!

**Pull request merged**